// @rxweb/reactive-forms/esm5/directives/form-control-directive.js
var CHECKBOX = 'checkbox';
var SELECT_MULTIPLE = "select-multiple";
var FormControlDirective = /** @class */ (function () {
    function FormControlDirective(element, formControl) {
        this.element = element;
        this.formControl = formControl;
        this.bindEvent();
        this.setValue();
        this.subscribeOnChange();
    }
    FormControlDirective.prototype.bindEvent = function () {
        this._onChange = this.onChange.bind(this);
        this._onBlur = this.onBlur.bind(this);
        this.element.addEventListener("input", this._onChange);
        this.element.addEventListener("blur", this._onBlur);
    };
    FormControlDirective.prototype.onChange = function (event) {
        this.formControl.setValue(event.target.value);
    };
    FormControlDirective.prototype.onBlur = function (event) {
        this.formControl.markAsTouched();
    };
    FormControlDirective.prototype.setValueOnControl = function () {
        this.element.value = this.formControl.value != null ? this.formControl.value : '';
    };
    FormControlDirective.prototype.setValue = function () {
        var value = this.formControl.value === "" || this.formControl.value === undefined ? "" : this.formControl.value;
        switch (this.element.type) {
            case CHECKBOX:
                this.element.checked = value !== null && value !== false;
                break;
            default:
                this.element.value = value;
                break;
        }
    };
    FormControlDirective.prototype.subscribeOnChange = function () {
        var _this = this;
        this.formControl.subscribe(function () {
            if (_this.element.value != _this.formControl.value)
                _this.setValue();
            if (_this.formControl.invalid) {
                if (!_this.element.classList.contains("rx-invalid"))
                    _this.element.classList.add("rx-invalid");
            }
            else {
                if (_this.element.classList.contains("rx-invalid"))
                    _this.element.classList.remove("rx-invalid");
            }
        });
    };
    FormControlDirective.prototype.destroy = function () {
        this.formControl.destroy();
        this.element.removeEventListener("input", this._onChange);
        this.element.removeEventListener("blur", this._onBlur);
    };
    return FormControlDirective;
}());
export { FormControlDirective };
