// @rxweb/reactive-forms/esm2015/util/reactive-form-config.js
export var ClientLibrary;
(function (ClientLibrary) {
    ClientLibrary[ClientLibrary["React"] = 0] = "React";
    ClientLibrary[ClientLibrary["Vue"] = 1] = "Vue";
    ClientLibrary[ClientLibrary["Angular"] = 2] = "Angular";
    ClientLibrary[ClientLibrary["Vanilla"] = 3] = "Vanilla";
    ClientLibrary[ClientLibrary["None"] = 4] = "None";
})(ClientLibrary || (ClientLibrary = {}));
export class ReactiveFormConfig {
    static set(jObject) {
        if (jObject)
            ReactiveFormConfig.json = jObject;
    }
    static get(path) {
        let jObject = undefined;
        if (ReactiveFormConfig.json) {
            let splitPath = path.split('.');
            for (let columnName of splitPath) {
                jObject = (!jObject) ? ReactiveFormConfig.json[columnName] : jObject[columnName];
                if (!jObject)
                    break;
            }
        }
        return jObject;
    }
}
ReactiveFormConfig.number = {};
ReactiveFormConfig.json = {};
ReactiveFormConfig.clientLib = ClientLibrary.None;
