// @rxweb/reactive-forms/esm5/core/form-group.state.js
import { FormGroupDirective } from "../directives/form-group.directive";
export var formGroupContainer = new (/** @class */ (function () {
    function class_1() {
        this.formGroups = {};
        this.groupKeyElements = {};
        this.formGroupDirectives = {};
    }
    class_1.prototype.get = function (groupKey) {
        return this.formGroups[groupKey];
    };
    class_1.prototype.saveFormGroup = function (formGroup) {
        var groupKey = formGroup.path;
        this.formGroups[formGroup.path] = formGroup;
        this.bindIt(groupKey);
    };
    class_1.prototype.mapElement = function (groupKey, element) {
        this.groupKeyElements[groupKey] = element;
        this.bindIt(groupKey);
    };
    class_1.prototype.bindIt = function (groupKey) {
        var _this = this;
        if (this.groupKeyElements[groupKey] && this.formGroups[groupKey]) {
            setTimeout(function () {
                _this.formGroupDirectives[groupKey] = new FormGroupDirective(_this.groupKeyElements[groupKey], _this.formGroups[groupKey]);
            }, 100);
        }
    };
    class_1.prototype.destroy = function (groupKey) {
        if (this.formGroupDirectives[groupKey]) {
            this.formGroupDirectives[groupKey].destroy();
            delete this.formGroupDirectives[groupKey];
        }
    };
    return class_1;
}()))();
