// @rxweb/reactive-forms/esm5/util/reactive-form-config.js
import * as tslib_1 from "tslib";
export var ClientLibrary;
(function (ClientLibrary) {
    ClientLibrary[ClientLibrary["React"] = 0] = "React";
    ClientLibrary[ClientLibrary["Vue"] = 1] = "Vue";
    ClientLibrary[ClientLibrary["Angular"] = 2] = "Angular";
    ClientLibrary[ClientLibrary["Vanilla"] = 3] = "Vanilla";
    ClientLibrary[ClientLibrary["None"] = 4] = "None";
})(ClientLibrary || (ClientLibrary = {}));
var ReactiveFormConfig = /** @class */ (function () {
    function ReactiveFormConfig() {
    }
    ReactiveFormConfig.set = function (jObject) {
        if (jObject)
            ReactiveFormConfig.json = jObject;
    };
    ReactiveFormConfig.get = function (path) {
        var e_1, _a;
        var jObject = undefined;
        if (ReactiveFormConfig.json) {
            var splitPath = path.split('.');
            try {
                for (var splitPath_1 = tslib_1.__values(splitPath), splitPath_1_1 = splitPath_1.next(); !splitPath_1_1.done; splitPath_1_1 = splitPath_1.next()) {
                    var columnName = splitPath_1_1.value;
                    jObject = (!jObject) ? ReactiveFormConfig.json[columnName] : jObject[columnName];
                    if (!jObject)
                        break;
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (splitPath_1_1 && !splitPath_1_1.done && (_a = splitPath_1.return)) _a.call(splitPath_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return jObject;
    };
    ReactiveFormConfig.number = {};
    ReactiveFormConfig.json = {};
    ReactiveFormConfig.clientLib = ClientLibrary.None;
    return ReactiveFormConfig;
}());
export { ReactiveFormConfig };
