// @rxweb/reactive-forms/esm2015/directives/form-control-directive.js
const CHECKBOX = 'checkbox';
const SELECT_MULTIPLE = "select-multiple";
export class FormControlDirective {
    constructor(element, formControl) {
        this.element = element;
        this.formControl = formControl;
        this.bindEvent();
        this.setValue();
        this.subscribeOnChange();
    }
    bindEvent() {
        this._onChange = this.onChange.bind(this);
        this._onBlur = this.onBlur.bind(this);
        this.element.addEventListener("input", this._onChange);
        this.element.addEventListener("blur", this._onBlur);
    }
    onChange(event) {
        this.formControl.setValue(event.target.value);
    }
    onBlur(event) {
        this.formControl.markAsTouched();
    }
    setValueOnControl() {
        this.element.value = this.formControl.value != null ? this.formControl.value : '';
    }
    setValue() {
        let value = this.formControl.value === "" || this.formControl.value === undefined ? "" : this.formControl.value;
        switch (this.element.type) {
            case CHECKBOX:
                this.element.checked = value !== null && value !== false;
                break;
            default:
                this.element.value = value;
                break;
        }
    }
    subscribeOnChange() {
        this.formControl.subscribe(() => {
            if (this.element.value != this.formControl.value)
                this.setValue();
            if (this.formControl.invalid) {
                if (!this.element.classList.contains("rx-invalid"))
                    this.element.classList.add("rx-invalid");
            }
            else {
                if (this.element.classList.contains("rx-invalid"))
                    this.element.classList.remove("rx-invalid");
            }
        });
    }
    destroy() {
        this.formControl.destroy();
        this.element.removeEventListener("input", this._onChange);
        this.element.removeEventListener("blur", this._onBlur);
    }
}
